### DISCUSSION TITLE
OTP based Login

### DISCUSSION THREAD

**Discussion by safvankothawala** [keycloak/keycloak#11762] [Q&A]
Hello,

I want to enable login in my application using following call flow:
Step 1: User visits the Application website and enters the Mobile Number or Username.
Step 2: Application requests Keycloak to generate the OTP for the user associated with the username. The OTP returned by Keycloak is sent to User via Email or SMS.
Step 3: User enters the OTP and the application sends the OTP to Keycloak to validate against that user.
Step 4: Keycloak validates the OTP and responds back with Access Token.

In the above call flow, I need 2 APIs from Keycloak.
1. API to generate the OTP for given username
2. API to validate the OTP provided by the User and respond back with Access Token

Please let me know if this login call flow is possible or not using the latest Keycloak version.

**Comment by abstractj** [comment]
@safvankothawala the scenario you described should be possible, by writing your own authentication provider. You can find an example here: https://github.com/dasniko/keycloak-2fa-sms-authenticator

I have one additional comment. The OTP was supposed to be an extra layer of security, not a replacement for passwords. Plus, it is strongly recommended to avoid sending OTPs via e-mail, or SMS. SMS is vulnerable to SIM Swaps, and should not be trusted at all.

**Reply by Nefcanto** [reply]
@abstractj, I read about SIM swaps and I understood it. But having a phone-OTP solution out of the box would benefit the whole community. Because our customers want it. Though it's less secure (and the attack requires personal action, which makes it even harder) it's much easier than username and password.

And I talk about my personal experience, asking users in a simple shop to turn on their 2FA won't work. They don't care.

When a giant like Telegram uses it, I can rest assured that it's secure enough for practical uses. And besides, by enabling this feature, Keycloak can still show a warning message to the admin in the admin console that this method is less secure because of all these attack possibilities.

**Reply by abstractj** [reply]
@Nefcanto if you're inclined to contribute towards this functionality, please consider opening a feature request in the project's upstream repository.  This allows other maintainers and the community to discuss its feasibility, security implications, and alignment with our goals. 

Your input on enhancing user experience while addressing security concerns is appreciated, and involving the community can help assess and refine your proposal.

**Reply by ak-verse** [reply]
The 2FA example you have shared is browser based authentication flow. In our app , we need to interact with Keycloak using REST API. How do we make sure keycloak session is maintained during the 2 step authentication process?
Requirement :
Using Direct Grant authentication flow for authenticating the users through keycloak, my app is registered as a confidential client

1. Implemented a custom SPI to generate an OTP and send to user (working fine). authentcate() fundtion sends SMS. 
2. Created a Authentication flow with 3 execution steps : username validation , password , custom authenticator SPI

Expectation: action() function from custom Authenticator class should be called after user enters otp through second REST API call and validate the OTP 

Problem:
When second API call is made with the execution ID, keycloak is failing to identify the paused session and link it as one authentication flow, getting an error saying password is null, invalid credentials meaning keycloak should be considering the second REST API call as a fresh auth request. Can you please help?

**Comment by codespearhead** [comment]
Can you please close this issue/discussion [1][2]?

This question has been asked several times before. You can find previous discussions by searching for “sms” or “email” in the GitHub search bar of the Keycloak repository [3][4].

Currently, there is no official support for using email or phone numbers for authentication in Keycloak. Therefore, any issues you encounter with third-party solutions should be reported in their respective repositories, not here.

If you're interested in this feature, feel free to share your use case here [5], and upvote it by adding a ⬆️ to the discussion’s description after reading this answer carefully [6].

***

[1] https://docs.github.com/en/issues/tracking-your-work-with-issues/administering-issues/closing-an-issue
[2] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#closing-a-discussion
[3] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=issues
[4] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=discussions
[5] https://github.com/keycloak/keycloak/discussions/27152
[6] https://github.com/keycloak/keycloak/discussions/27152#discussioncomment-10923758
